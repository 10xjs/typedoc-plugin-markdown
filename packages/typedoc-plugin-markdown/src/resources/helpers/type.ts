import { DeclarationReflection, SignatureReflection } from 'typedoc';
import {
  ArrayType,
  IntersectionType,
  IntrinsicType,
  QueryType,
  ReferenceType,
  ReflectionType,
  StringLiteralType,
  TupleType,
  TypeOperatorType,
  TypeParameterType,
  UnionType,
} from 'typedoc/dist/lib/models/types';

import MarkdownTheme from '../../theme';
import { escape } from './escape';

export function type(
  this:
    | ArrayType
    | IntersectionType
    | IntrinsicType
    | ReferenceType
    | StringLiteralType
    | TupleType
    | UnionType
    | TypeOperatorType
    | TypeParameterType
    | QueryType,
  expandType = false,
) {
  if (
    this instanceof ReferenceType &&
    (this.reflection || (this.name && this.typeArguments))
  ) {
    return getReferenceType(this);
  }

  if (this instanceof ArrayType && this.elementType) {
    return getArrayType(this);
  }

  if (this instanceof UnionType && this.types) {
    return getUnionType(this);
  }

  if (this instanceof IntersectionType && this.types) {
    return getIntersectionType(this);
  }

  if (this instanceof TupleType && this.elements) {
    return getTupleType(this);
  }

  if (this instanceof IntrinsicType && this.name) {
    return getIntrinsicType(this);
  }

  if (this instanceof StringLiteralType && this.value) {
    return getStringLiteralType(this);
  }

  if (
    this instanceof ReflectionType &&
    (this.declaration.children || this.declaration.indexSignature)
  ) {
    return expandType ? getLiteralType(this.declaration) : 'object';
  }

  if (this instanceof DeclarationReflection && this.children) {
    return expandType ? getLiteralType(this) : 'object';
  }

  if (this instanceof ReflectionType && this.declaration.signatures) {
    return expandType
      ? getFunctionType(this.declaration.signatures)
      : 'function';
  }

  if (this instanceof DeclarationReflection && this.signatures) {
    return expandType ? getFunctionType(this.signatures) : 'function';
  }

  if (this instanceof TypeOperatorType) {
    return getTypeOperatorType(this);
  }

  if (this instanceof QueryType) {
    return getQueryType(this);
  }

  return this ? escape(this.toString()) : '';
}

function getReferenceType(model: ReferenceType) {
  const reflection =
    model.reflection && model.reflection.url
      ? [
          `[${
            model.reflection.name
          }](${MarkdownTheme.HANDLEBARS.helpers.relativeURL(
            model.reflection.url,
          )})`,
        ]
      : [model.name];
  if (model.typeArguments && model.typeArguments.length > 0) {
    reflection.push(
      `\\<${model.typeArguments
        .map((typeArgument) => `${type.call(typeArgument)}`)
        .join(', ')}>`,
    );
  }
  return reflection.join('');
}

function getArrayType(model: ArrayType) {
  const arrayType = type.call(model.elementType);
  return model.elementType.type === 'union'
    ? `(${arrayType})[]`
    : `${arrayType}[]`;
}

function getUnionType(model: UnionType) {
  return model.types
    .map((unionType) => type.call(unionType, true))
    .join(' \\| ');
}

function getIntersectionType(model: IntersectionType) {
  return model.types
    .map((intersectionType) => type.call(intersectionType))
    .join(' & ');
}

function getTupleType(model: TupleType) {
  return `[${model.elements.map((element) => type.call(element)).join(', ')}]`;
}

function getIntrinsicType(model: IntrinsicType) {
  return model.name;
}

function getStringLiteralType(model: StringLiteralType) {
  return `\\"${escape(model.value)}\\"`;
}

function getLiteralType(model: DeclarationReflection) {
  let indexSignature = '';
  const declarationIndexSignature = model.indexSignature;
  if (declarationIndexSignature) {
    const key = declarationIndexSignature.parameters
      ? declarationIndexSignature.parameters.map(
          (param) => `[${param.name}:${param.type}]`,
        )
      : '';
    const obj = type.call(declarationIndexSignature.type, true);
    indexSignature = `${key}: ${obj}; `;
  }
  let types;
  if (model.children) {
    types = model.children.map((obj) => {
      return `${obj.name}${obj.flags.isOptional ? '?' : ''}: ${type.call(
        obj.signatures || obj.children ? obj : obj.type,
        true,
      )} ${obj.defaultValue ? `= ${escape(obj.defaultValue)}` : ''}`;
    });
  }
  return `{ ${indexSignature ? indexSignature : ''}${
    types ? types.join('; ') : ''
  } }${model.defaultValue ? `= ${escape(model.defaultValue)}` : ''}`;
}

export function getFunctionType(modelSignatures: SignatureReflection[]) {
  const functions = modelSignatures.map((fn) => {
    const typeParams = fn.typeParameters
      ? `\\<${fn.typeParameters
          .map((typeParameter) => typeParameter.name)
          .join(', ')}>`
      : [];
    const params = fn.parameters
      ? fn.parameters.map((param) => {
          return `${param.flags.isRest ? '...' : ''}${param.name}${
            param.flags.isOptional ? '?' : ''
          }: ${type.call(param.type ? param.type : param)}`;
        })
      : [];
    const returns = type.call(fn.type, true);

    return typeParams + `(${params.join(',')}) => ${returns}`;
  });
  return functions.join('');
}

function getTypeOperatorType(model: TypeOperatorType) {
  return `${model.operator} ${type.call(model.target)}`;
}

function getQueryType(model: QueryType) {
  return `typeof ${type.call(model.queryType)}`;
}
