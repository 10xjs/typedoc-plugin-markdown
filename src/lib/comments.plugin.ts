import * as FS from 'fs-extra';
import * as Handlebars from 'handlebars';
import * as Path from 'path';
import { MarkedLinksPlugin, Reflection } from 'typedoc';
import { Component, ContextAwareRendererComponent } from 'typedoc/dist/lib/output/components';
import { RendererEvent } from 'typedoc/dist/lib/output/events';

@Component({ name: 'comments' })
export class CommentsPlugin extends ContextAwareRendererComponent {
  /**
   * The path referenced files are located in.
   */
  private includes?: string;

  /**
   * Path to the output media directory.
   */
  private mediaDirectory?: string;

  /**
   * The pattern used to find references in markdown.
   */
  private includePattern: RegExp = /\[\[include:([^\]]+?)\]\]/g;

  /**
   * The pattern used to find media links.
   */
  private mediaPattern: RegExp = /media:\/\/([^ "\)\]\}]+)/g;

  /**
   * Regular expression for detecting bracket links.
   */
  private brackets: RegExp = /\[\[([^\]]+)\]\]/g;

  /**
   * Regular expression for detecting inline tags like {@link ...}.
   */
  private inlineTag: RegExp = /(?:\[(.+?)\])?\{@(link|linkcode|linkplain)\s+((?:.|\n)+?)\}/gi;

  listInvalidSymbolLinks: boolean;

  private warnings: string[] = [];

  /**
   * Create a new MarkedPlugin instance.
   */
  initialize() {
    super.initialize();

    this.listenTo(
      this.owner,
      {
        [RendererEvent.END]: this.onEndRenderer,
      },
      undefined,
      100,
    );

    const that = this;

    Handlebars.registerHelper('comment', function(this: string) {
      return that.parseMarkdown(this);
    });
  }

  /**
   * Parse the given markdown string and return the resulting html.
   *
   * @param text  The markdown string that should be parsed.
   * @param context  The current handlebars context.
   * @returns The resulting html string.
   */
  public parseMarkdown(text: string) {
    const context = Object.assign(text, '');

    this.includes = this.application.options.getValue('includes');
    this.mediaDirectory = this.application.options.getValue('media');
    this.listInvalidSymbolLinks = this.application.options.getValue('listInvalidSymbolLinks');

    if (this.includes) {
      text = text.replace(this.includePattern, (match: string, path: string) => {
        path = Path.join(this.includes!, path.trim());
        if (FS.existsSync(path) && FS.statSync(path).isFile()) {
          const contents = FS.readFileSync(path, 'utf-8');
          if (path.substr(-4).toLocaleLowerCase() === '.hbs') {
            const template = Handlebars.compile(contents);
            return template(context);
          } else {
            return contents;
          }
        } else {
          return '';
        }
      });
    }

    if (this.mediaDirectory) {
      text = text.replace(this.mediaPattern, (match: string, path: string) => {
        if (FS.existsSync(Path.join(this.mediaDirectory!, path))) {
          return this.getRelativeUrl('media') + '/' + path;
        } else {
          return match;
        }
      });
    }

    return this.replaceInlineTags(this.replaceBrackets(text));
  }

  /**
   * Find all references to symbols within the given text and transform them into a link.
   *
   * This function is aware of the current context and will try to find the symbol within the
   * current reflection. It will walk up the reflection chain till the symbol is found or the
   * root reflection is reached. As a last resort the function will search the entire project
   * for the given symbol.
   *
   * @param text  The text that should be parsed.
   * @returns The text with symbol references replaced by links.
   */
  private replaceBrackets(text: string): string {
    return text.replace(this.brackets, (match: string, content: string): string => {
      const split = MarkedLinksPlugin.splitLinkText(content);
      return this.buildLink(match, split.target, split.caption);
    });
  }

  /**
   * Find symbol {@link ...} strings in text and turn into html links
   *
   * @param text  The string in which to replace the inline tags.
   * @return      The updated string.
   */
  private replaceInlineTags(text: string): string {
    return text.replace(this.inlineTag, (match: string, leading: string, tagName: string, content: string): string => {
      const split = MarkedLinksPlugin.splitLinkText(content);
      const target = split.target;
      const caption = leading || split.caption;
      const monospace = tagName === 'linkcode';

      return this.buildLink(match, target, caption, monospace);
    });
  }

  /**
   * Format a link with the given text and target.
   *
   * @param original   The original link string, will be returned if the target cannot be resolved..
   * @param target     The link target.
   * @param caption    The caption of the link.
   * @param monospace  Whether to use monospace formatting or not.
   * @returns A html link tag.
   */
  private buildLink(original: string, target: string, caption: string, monospace?: boolean): string {
    let reflection: Reflection | undefined;

    if (this.reflection) {
      reflection = this.reflection.findReflectionByName(target);
    } else if (this.project) {
      reflection = this.project.findReflectionByName(target);
    }

    if (reflection && reflection.url) {
      if (this.urlPrefix.test(reflection.url)) {
        target = reflection.url;
      } else {
        target = this.getRelativeUrl(reflection.url);
      }
    } else {
      const fullName = (this.reflection || this.project)!.getFullName();
      this.warnings.push(`In ${fullName}: ${original}`);
      return original;
    }

    if (monospace) {
      caption = '`' + caption + '`';
    }

    return `[${caption}](${target})`;
  }

  /**
   * Triggered when [[Renderer]] is finished
   */
  onEndRenderer(event: RendererEvent) {
    if (this.listInvalidSymbolLinks && this.warnings.length > 0) {
      this.application.logger.write('');
      this.application.logger.warn(
        'Found invalid symbol reference(s) in JSDocs, ' +
          'they will not render as links in the generated documentation.',
      );

      for (const warning of this.warnings) {
        this.application.logger.write('  ' + warning);
      }
    }
  }
}
