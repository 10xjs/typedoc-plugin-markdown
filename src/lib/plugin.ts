import * as path from 'path';
import { Application, DeclarationReflection, ProjectReflection } from 'typedoc';
import { Converter } from 'typedoc/dist/lib/converter';
import { Component, ConverterComponent } from 'typedoc/dist/lib/converter/components';
import { Reflection } from 'typedoc/dist/lib/models/reflections/abstract';
import { PageEvent, RendererEvent } from 'typedoc/dist/lib/output/events';
import { Renderer } from 'typedoc/dist/lib/output/renderer';
import { Options } from 'typedoc/dist/lib/utils/options';

import { MarkdownTheme } from './theme/theme';
import { BitbucketTheme } from './theme/theme.bitbucket';
import { DocusaurusTheme } from './theme/theme.docusaurus';
import { GitbookTheme } from './theme/theme.gitbook';

@Component({ name: 'markdown' })
export class MarkdownPlugin extends ConverterComponent {
  static theme: MarkdownTheme | DocusaurusTheme;
  static disableOutputCheck: boolean;
  static application: Application;
  static reflection: Reflection;
  static location: string;
  static page: PageEvent;
  static settings: {
    platform?: 'gitbook' | 'ducusaurus' | 'bitbucket';
    hideSources?: boolean;
    readme?: string;
    includes?: string;
    media?: string;
    mode?: string;
    mdHideSources?: boolean;
    mdSourceRepo?: string;
  };
  static project: ProjectReflection;

  initialize() {
    this.listenTo(this.owner, {
      [Converter.EVENT_RESOLVE_BEGIN]: this.onBegin,
    });
    this.listenTo(this.application.renderer, {
      [PageEvent.BEGIN]: this.onPageBegin,
      [PageEvent.END]: this.onPageEnd,
      [RendererEvent.BEGIN]: this.onRenderBegin,
    });
  }

  onBegin() {
    MarkdownPlugin.application = this.application;
    MarkdownPlugin.settings = this.application.options.getRawValues();
    MarkdownPlugin.setTheme(this.application.renderer, this.application.options);
  }

  onRenderBegin(renderer: RendererEvent) {
    if (MarkdownPlugin.theme) {
      MarkdownPlugin.project = renderer.project;
    }
  }

  onPageBegin(page: PageEvent) {
    if (MarkdownPlugin.theme) {
      MarkdownPlugin.reflection = page.model instanceof DeclarationReflection ? page.model : undefined;
      MarkdownPlugin.location = page.model.url ? page.model.url : '';
    }
  }

  onPageEnd(page: PageEvent) {
    if (MarkdownPlugin.theme) {
      page.contents = page.contents ? MarkdownPlugin.formatContents(page.contents) : '';
    }
  }

  static setTheme(renderer: Renderer, options: Options) {
    const themeName = options.getValue('theme');
    const themePath = path.join(__dirname, './theme/');
    const media = options.getValue('media');
    const platform = options.getValue('platform') || options.getValue('mdEngine');
    const theme = themeName === 'markdown' ? this.getTheme(platform, renderer, themePath, options) : null;
    if (theme) {
      if (media && (theme instanceof DocusaurusTheme || theme instanceof GitbookTheme)) {
        MarkdownPlugin.application.logger.warn(`[typedoc-markdown-plugin] media option is currently not supported in ${platform} theme`);
        options.setValue('media', null);
      }
      renderer.theme = renderer.addComponent('theme', theme);
      MarkdownPlugin.theme = theme;
    } else {
      MarkdownPlugin.application.logger.write('Markdown theme not set');
    }
  }

  static getTheme(platform: string, renderer: Renderer, themePath: string, options: any) {
    if (platform) {
      if (platform === 'gitbook') {
        return new GitbookTheme(renderer, themePath, options);
      }
      if (platform === 'docusaurus') {
        return new DocusaurusTheme(renderer, themePath, options);
      }
      if (platform === 'bitbucket') {
        return new BitbucketTheme(renderer, themePath, options);
      }
    }
    return new MarkdownTheme(renderer, themePath, options);
  }

  static formatContents(contents: string) {
    return contents
      .replace(/^ +/gm, '')
      .replace(/[\r\n]{3,}/g, '\n\n')
      .replace(/!spaces/g, '')
      .trim();
  }
}
