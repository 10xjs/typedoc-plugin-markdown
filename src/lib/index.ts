import { Application } from 'typedoc/dist/lib/application';
import { ParameterType } from 'typedoc/dist/lib/utils/options/declaration';
import { MarkdownPlugin } from './plugin';

export = (PluginHost: Application) => {
  const app = PluginHost.owner;
  if (app.converter.hasComponent('markdown')) {
    return;
  }

  app.options.addDeclaration({
    component: 'markdown',
    help: 'Markdown Plugin: (docusaurus|gitbook|bitbucket) Specifies the markdown target.',
    name: 'platform',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    component: 'markdown',
    help: 'Markdown Plugin: Suppress file sources from output.',
    name: 'hideSources',
    type: ParameterType.Boolean,
  });

  app.options.addDeclaration({
    component: 'markdown',
    help: 'Markdown Plugin: Deprectated - use --platform.',
    name: 'mdEngine',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    component: 'markdown',
    help: 'Markdown Plugin: Deprectated - use --hideSources.',
    name: 'mdHideSources',
    type: ParameterType.Boolean,
  });

  app.converter.addComponent('markdown', new MarkdownPlugin(app.converter));
};
