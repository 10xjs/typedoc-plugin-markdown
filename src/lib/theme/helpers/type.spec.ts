import { IntrinsicType } from 'typedoc/dist/lib/models';
import { type } from './type';

describe(`type helper`, () => {
  test(`should compile intrinsic type`, () => {
    const intrinsicType = new IntrinsicType('string');
    const result = type.call(intrinsicType);
    expect(result).toMatchSnapshot();
  });
});
