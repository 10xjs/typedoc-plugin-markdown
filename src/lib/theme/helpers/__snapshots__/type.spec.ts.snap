// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type helper should compile intrinsic type 1`] = `"string"`;
